### PR TITLE
Fix #117: Tag pull requests with build older than 2 days

### LINKED ISSUE #117: Oppiabot should tag PRs with builds passing but last commit > 2 days ago
Oppiabot should tag PRs with builds passing but last commit > 2 days ago, with a warning message / label saying “build might be stale, please update from develop”. Then reviewers/authors can take a closer look, and if someone introduces a new check, they can look at PRs with this label and restart the checks if it makes sense.

### PR DESCRIPTION
**Pull request by jameesjohn**
## Explanation
Fixes #117: Tags all pull request with build older than 2 days by pinging the author to merge from develop and adding the `PR: don't merge - STALE BUILD` label to the pull request.
The features added here have been tested at https://github.com/jameesjohn/oppia/pull/39#issuecomment-717203672.

## Checklist
- [x] I have successfully deployed my own instance of Oppiabot.
  - You can find instructions for doing this [here](https://github.com/oppia/oppiabot/wiki/Deploying-your-own-instance-of-the-oppiabot).
- [x] I have manually tested all the changes made in this PR following the [manual tests matrix](https://github.com/oppia/oppiabot/wiki/Manual-Tests-Matrix).

### CODE REVIEW

File: lib/periodicChecks.js [hunk 1]
```diff
@@ -174,7 +175,58 @@ const ensureAllIssuesHaveProjects = async (context) => {
...
+
+        // Ping PR author.
+        const commentParams = context.repo({
+          issue_number: pullRequest.number,
+          body:
+            'Hi @' + pullRequest.user.login + ', the build of this PR is ' +
+            'stale and this could result in tests failing in develop. Please' +
+            ' update this pull request with the latest changes from develop,' +
+            ' Thanks.',
+        });
```
Comment (vojtechjelinek) [inline lib/periodicChecks.js:221]:
```suggestion
            'Hi @' + pullRequest.user.login + ', the build of this PR is ' +
            'stale and this could result in tests failing in develop. Please ' +
            'update this pull request with the latest changes from develop, ' +
            'Thanks.',
        });
```

File: constants.js [hunk 2]
```diff
@@ -108,6 +108,9 @@ const checksWhitelist = {
     [issuesAssignedEvent]: [],
     [pushEvent]: []
   },
+  'comment-on-pr': {
+    [periodicCheckEvent]: [periodicCheck],
+  }
```
Comment (vojtechjelinek) [inline constants.js:113]: Why is this needed?

File: lib/utils.js [hunk 3]
```diff
@@ -29,6 +29,8 @@ const CODE_OWNERS_FILE_URL =
 const CHANGELOG_START = 'PR CHANGELOG';
 const SUCCESS_STATUS = 204;
 const THREE_MINUTES = 60 * 1000 * 3;
+let MIN_BUILD_DATE = new Date();
+MIN_BUILD_DATE.setDate(MIN_BUILD_DATE.getDate() - 2); // 2 days prior.
```
Comment (vojtechjelinek) [inline lib/utils.js:33]: I would just set the two days here, and count the exact date in `checkAndTagPRsWithOldBuilds`.

File: spec/periodicChecksSpec.js [hunk 4]
```diff
@@ -825,4 +913,205 @@ describe('Periodic Checks Module', () => {
       });
     });
   });
+
+  describe('When pull request has an old build', () => {
```
Comment (vojtechjelinek) [inline spec/periodicChecksSpec.js:917]:
```suggestion
  describe('when pull request has an old build', () => {
```

**Review by vojtechjelinek — CHANGES_REQUESTED** [review]
Thanks! Looks pretty solid, just a few comments.

File: lib/periodicChecks.js [hunk 5]
```diff
@@ -174,7 +175,58 @@ const ensureAllIssuesHaveProjects = async (context) => {
...
+  allOpenPullRequests.forEach(async (pullRequest) => {
+    const containsOldBuildLabel = pullRequest.labels.some(
+      label => label.name === OLD_BUILD_LABEL
+    );
+    // If label has already been added, do nothing.
+    if (!containsOldBuildLabel) {
+    // Get last commit for pull request.
+    const lastCommitSha = pullRequest.head.sha;
+
+    const pRCommits = await context.github.pulls.listCommits(
```
[the remark this answers is not in the document]
Comment (jameesjohn) [reply lib/periodicChecks.js]: Yeah
Updated, thanks!

File: lib/periodicChecks.js (hunk 1 above)
Comment (jameesjohn) [reply lib/periodicChecks.js:220]: Adding these will make those lines 81 characters long.

File: lib/utils.js (hunk 3 above)
Comment (jameesjohn) [reply lib/utils.js:33]: I don't think I understand what you mean here.
Also, I'm setting it here so I will be able to mock it while testing.

File: spec/periodicChecksSpec.js (hunk 4 above)
Comment (jameesjohn) [reply spec/periodicChecksSpec.js]: Done.

File: lib/utils.js (hunk 3 above)
Comment (vojtechjelinek) [reply lib/utils.js:33]: Oh, okay, fine then.

File: lib/periodicChecks.js [hunk 6]
```diff
@@ -174,7 +175,57 @@ const ensureAllIssuesHaveProjects = async (context) => {
...
+            'Hi @' + pullRequest.user.login + ', the build of this PR is ' +
+            'stale and this could result in tests failing in develop. Please' +
+            ' update this pull request with the latest changes from develop,' +
+            ' Thanks.',
+        });
+        context.github.issues.createComment(commentParams);
+      }
+    }
+  );
+};
```
Comment (vojtechjelinek) [inline lib/periodicChecks.js:225]:
```suggestion
    const lastCommitDate = new Date(lastCommit.commit.author.date);
    // Commit is older than 2 days
    if (utilityModule.MIN_BUILD_DATE > lastCommitDate) {
      const labelParams = context.repo({
        issue_number: pullRequest.number,
        labels: [OLD_BUILD_LABEL],
      });
      context.github.issues.addLabels(labelParams);

      // Ping PR author.
      const commentParams = context.repo({
        issue_number: pullRequest.number,
        body:
          'Hi @' + pullRequest.user.login + ', the build of this PR is ' +
          'stale and this could result in tests failing in develop. Please ' +
          'update this pull request with the latest changes from develop, ' +
          'Thanks.',
        });
      context.github.issues.createComment(commentParams);
    }
  });
};
```

Comment (vojtechjelinek) [reply lib/periodicChecks.js:225]: You added two spaces indentation on line 204 without any reason.

File: lib/periodicChecks.js [hunk 7]
```diff
@@ -26,6 +26,7 @@ const {
   getAllProjectCards,
   hasPendingReviews
 } = require('./utils');
+const utilityModule = require('./utils');
```
Comment (vojtechjelinek) [inline lib/periodicChecks.js:29]: Can you include the `MIN_BUILD_DATE` in the import aboce this?

File: spec/periodicChecksSpec.js [hunk 8]
```diff
@@ -825,4 +913,206 @@ describe('Periodic Checks Module', () => {
...
+                'https://api.github.com/repos/jameesjohn/oppia/commits/525d6058a626b46154ed3731141a95e70ebb0aef',
+              html_url:
+                'https://github.com/jameesjohn/oppia/commit/525d6058a626b46154ed3731141a95e70ebb0aef',
+              comments_url:
+                'https://api.github.com/repos/jameesjohn/oppia/commits/525d6058a626b46154ed3731141a95e70ebb0aef/comments',
+              author: [Object],
+              committer: [Object],
+              parents: [Array],
+            },
+          };
```
Comment (vojtechjelinek) [inline spec/periodicChecksSpec.js:1022]: Could you extract the commit data into separate variables?

File: lib/periodicChecks.js [hunk 9]
```diff
@@ -174,7 +175,57 @@ const ensureAllIssuesHaveProjects = async (context) => {
   });
 };
 
+/**
+ * This function comments on and tags pull requests in which the last commit was created 2 days ago
+ * @param {import('probot').Context} context
+ */
+const checkAndTagPRsWithOldBuilds = async (context) => {
+  const OLD_BUILD_LABEL = "PR: don't merge - STALE BUILD";
+  console.log({MIN_BUILD_DATE: utilityModule.MIN_BUILD_DATE});
```
Comment (vojtechjelinek) [inline lib/periodicChecks.js:184]: We can probably remove this?

File: lib/periodicChecks.js (hunk 1 above)
Comment (vojtechjelinek) [reply lib/periodicChecks.js:220]: Included in a different review comment.

**Review by vojtechjelinek — CHANGES_REQUESTED** [review]
Thanks! Just a few remaining comments.

File: lib/periodicChecks.js (hunk 7 above)
Comment (jameesjohn) [reply lib/periodicChecks.js:29]: The issue with that is I won't be able to mock it in the tests.

File: lib/periodicChecks.js (hunk 9 above)
Comment (jameesjohn) [reply lib/periodicChecks.js]: Done.

File: lib/periodicChecks.js (hunk 6 above)
Comment (jameesjohn) [reply lib/periodicChecks.js]: Modified

File: spec/periodicChecksSpec.js (hunk 8 above)
Comment (jameesjohn) [reply spec/periodicChecksSpec.js]: Done.

File: lib/periodicChecks.js (hunk 7 above)
Comment (vojtechjelinek) [reply lib/periodicChecks.js:29]: Ah, I see, okay.

File: lib/periodicChecks.js [hunk 10]
```diff
@@ -174,7 +175,57 @@ const ensureAllIssuesHaveProjects = async (context) => {
...
+      context.github.issues.addLabels(labelParams);
+
+      // Ping PR author.
+      const commentParams = context.repo({
+        issue_number: pullRequest.number,
+        body:
+            'Hi @' + pullRequest.user.login + ', the build of this PR is ' +
+            'stale and this could result in tests failing in develop. Please' +
+            ' update this pull request with the latest changes from develop,' +
+            ' Thanks.',
```
Comment (vojtechjelinek) [inline lib/periodicChecks.js:219]: Now that everything is shifted you should be able to do 
```suggestion
            'stale and this could result in tests failing in develop. Please' +
            'update this pull request with the latest changes from develop, ' +
            'Thanks. ',
```

**Review by vojtechjelinek — APPROVED** [review]
LGTM! Just one small nit.

File: lib/periodicChecks.js (hunk 10 above)
Comment (jameesjohn) [reply lib/periodicChecks.js]: Done.

**Comment by vojtechjelinek** [comment]
@jameesjohn The tests are failing?
